### PR TITLE
feat: Add Llama.cpp Generator

### PR DESCRIPTION
**Pull request by awinml**
### Related Issues:

Migrated from: https://github.com/awinml/llama-cpp-haystack/tree/main

fixes parts of #177 

### Proposed Changes

Introduces `LlamaCppGenerator` to load models quantized using llama.cpp (GGUF) for text generation.

- Examples have been added to the `examples` directory and README, including a standalone example and one that shows usage in a RAG pipeline.
- A new label `llama_cpp` has been added to the [labeler.yml](https://github.com/deepset-ai/haystack-core-integrations/blob/main/.github/labeler.yml).
- A Github workflow (`llama_cpp.yml`) has been added to run the tests suite on the CI.

### Design

```python
@component
class LlamaCppGenerator:
    def __init__(
        self,
        model_path: str,
        n_ctx: Optional[int] = 0,
        n_batch: Optional[int] = 512,
        model_kwargs: Optional[Dict[str, Any]] = None,
        generation_kwargs: Optional[Dict[str, Any]] = None,
    ):
    ...

    def warm_up(self):
        if self.model is None:
            self.model = Llama(**self.model_kwargs)

    @component.output_types(replies=List[str], meta=List[Dict[str, Any]])
    def run(self, prompt: str, generation_kwargs: Optional[Dict[str, Any]] = None):
        ...
        return {"replies": replies, "meta": [output]}
```

- The `warm_up()` method stores the loaded model config in `self.model`. Llama.cpp loads the model into memory during the generation call, so an explicit backend is not required.

- The `model_path`, `n_ctx`, `n_batch` arguments have been exposed for convenience and can be directly passed to the Generator during initialization as keyword arguments.  

- The `model_kwargs` parameter can be used to pass additional arguments when initializing the model. In case of duplication, these kwargs override `model_path`, `n_ctx`, and `n_batch` init parameters.

- The `generation_kwargs` parameter can be used to pass additional generation arguments like `max_tokens`, `temperature`, `top_k`, `top_p`, etc to the model during inference. 

- The `generation_kwargs` can also be passed to the `run` method of the generator directly.

See Llama.cpp's [model documentation](https://llama-cpp-python.readthedocs.io/en/latest/api-reference/#llama_cpp.Llama.__init__) and [`create_completion` documentation](https://llama-cpp-python.readthedocs.io/en/latest/api-reference/#llama_cpp.Llama.create_completion) for more information on the available model arguments and generation parameters.

### Notes for reviewer

The [OpenChat GGUF](https://huggingface.co/TheBloke/openchat-3.5-1210-GGUF/resolve/main/openchat-3.5-1210.Q3_K_S.gguf) is downloaded in the `tests/models` directory and used for the integration tests. The file size of the model is ~3.5 GB.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -73,6 +73,7 @@ deepset-haystack
...
+| [ollama-haystack](integrations/ollama/)                                         | Generator           | [![PyPI - Version](https://img.shields.io/pypi/v/ollama-haystack.svg?color=orange)](https://pypi.org/project/ollama-haystack)                            | [![Test / ollama](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/ollama.yml/badge.svg)](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/ollama.yml)                                                               |
+| [llama-cpp-haystack](integrations/llama_cpp/)                                   | Generator           | [![PyPI - Version](https://img.shields.io/pypi/v/ollama-haystack.svg?color=orange)](https://pypi.org/project/llama-cpp-haystack)                         | [![Test / llama-cpp](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/llama_cpp.yml/badge.svg)](https://github.com/deepset-ai/haystack-core-integrations/actions/workflows/llama_cpp.yml)                                                      |
```
Comment (masci) [inline README.md:79]: nit: can we put this entry in alphabetical order?

File: integrations/llama_cpp/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    "integration: marks tests as slow (deselect with '-m \"not integration\"')",
+]
+addopts = [
+    "--import-mode=importlib",
+]
+
+[[tool.mypy.overrides]]
+module = [
+    "haystack.*",
+    "pytest.*"
```
Comment (masci) [inline integrations/llama_cpp/pyproject.toml:182]: The CI is failing on mypy, I think you need to add `llama_cpp` here

**Review by masci — CHANGES_REQUESTED** [review]
Thanks for this contribution! I left a couple of comments but it looks good.

File: README.md (hunk 1 above)
Comment (awinml) [reply README.md]: Fixed, reordered some of the other entries as well.

**Review by masci — APPROVED** [review]
🚀

**Pull request merged**